test/library/gitlabidp.go

```diff
@@ -0,0 +1,316 @@
+	require.NoError(t, err)
+
+	configclients, err := configv1client.NewForConfig(kubeconfig)
+	require.NoError(t, err)
+
+	nsName, gitlabHost, cleanup := deployPod(t, kubeClients, routeClient,
+		"gitlab",
+		"docker.io/gitlab/gitlab-ce:latest",
+		[]corev1.EnvVar{
+			// configure password for GitLab root user
+			{Name: "GITLAB_OMNIBUS_CONFIG", Value: "gitlab_rails['initial_root_password']='password';"},
+		},
+		[]corev1.Volume{
+			{
+				Name: "configdir",
+				VolumeSource: corev1.VolumeSource{
+					EmptyDir: &corev1.EmptyDirVolumeSource{},
+				},
+			},
+		},
+		[]corev1.VolumeMount{
+			{
+				Name:      "configdir",
+				SubPath:   "gitlab/config",
+				MountPath: "/etc/gitlab",
+			},
+			{
+				Name:      "configdir",
+				SubPath:   "gitlab/logs",
+				MountPath: "/var/log/gitlab",
+			},
+			{
+				Name:      "configdir",
+				SubPath:   "gitlab/data",
+				MountPath: "/var/opt/gitlab",
+			},
+		},
+		corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				// it's a bit resource-heavy, it will probably eat more during start-up
+				// max observed cpu: 1420m
+				// max observed memory: 2001Mi
+				// but that's more than our devel clusters provide, let's just request some minimum
+				// and have the pod eat more
+				"cpu":    resource.MustParse("500m"),
+				"memory": resource.MustParse("1500Mi"),
+			},
+		},
+	)
+	cleanups = []func(){cleanup}
+
+	gitlabURL := "https://" + gitlabHost
+	transport, err := rest.TransportFor(kubeconfig)
+	defer func() {
+		if err != nil {
+			for _, c := range cleanups {
+				c()
+			}
+		}
+	}()
+	require.NoError(t, err)
+
+	gitlabClient := GitLabClientFor(t, transport, gitlabURL)
+
+	// TODO: future us might try to configure whitelist IPs so that we can configure
+	// health and readiness probes via gitlab's /-/health and /-/readiness endpoints
+	WaitForHTTPStatus(t, 10*time.Minute, gitlabClient.client, gitlabClient.gitlabURL.String(), 200)
+	err = gitlabClient.PasswordAuthenticate("root", "password")
+	require.NoError(t, err, "failed to login as root")
+
+	openshiftIDPName := fmt.Sprintf("gitlab-test-%s", nsName)
+
+	app, err := gitlabClient.CreateApplication(
+		"openshift-non-confidential",
+		[]string{fmt.Sprintf("https://oauth.server.ip/oauth2callback/%s", openshiftIDPName)}, // FIXME: a real redirect URI
+		[]string{"openid", "profile", "email"},
+		false)
+
+	appID := app["application_id"].(string)
+	appSecret := app["secret"].(string)
+
+	_, err = kubeClients.CoreV1().Secrets("openshift-config").Create(context.TODO(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "gitlab-secret",
+				Labels: CAOE2ETestLabels(),
+			},
+			Data: map[string][]byte{
+				"clientSecret": []byte(appSecret),
+			},
+		},
+		metav1.CreateOptions{},
+	)
+	require.NoError(t, err, "failed to create gitlab client secret")
+	cleanups = append(cleanups, func() {
+		kubeClients.CoreV1().Secrets("openshift-config").Delete(context.TODO(), "gitlab-secret", metav1.DeleteOptions{})
```

Comment:
Log on failure?

Separately, what are your thoughts on introducing a static checker to catch these kinds of mechanical errors (e.g. implicitly ignored return values)?

Reply:
Will add the logging.

I haven't used static checkers in golang too much, outside of what my IDE provides at least, I think that they make much more sense for scripting languages, whereas go with its static typing and strong pressure on the code structure may not need it that much, and could be more of a nuisance than help (mostly with adding error ignoring comments).

Reply:
While I share your concern at poorly configured tools causing problems, I've seen some positive outcomes. Most recently in openshift/kubernetes, enabling upstream static analysis checks that we previously had disabled revealed a number of mechanical gotchas (similar to the one here) in our persistent carries. A good tool supports disabling checks that are not useful for a codebase. I think at a minimum ensuring that return values need to be explicitly ignored (e.g. `_, _ := myfunc()` vs `myfunc()`) would be worth enabling since it is such a common cause of accidentally ignoring errors.